Simulator/actions/messaging_actions.js

```diff
@@ -0,0 +1,257 @@
+    // using the schema info
+    const { wdcShouldFetchAllTables } = getState();
+    if (validateSchema(schema)) {
+      const newTables = schema.reduce((tables, tableInfo) => {
+        const newTable = {
+          schema: tableInfo,
+          data: [],
+        };
+
+        return { ...tables, [tableInfo.id]: newTable };
+      }, {});
+      dispatch(simulatorActions.addTables(newTables));
+
+      // if we are supposed to automatically fetch the tables, do it
+      if (wdcShouldFetchAllTables) {
+        dispatch(fetchAllData());
+      } else {
+        // We we aren't going to fetch immediately
+        // consider the phase paused until the getGeta
+        // message is sent. This is so we can properly
+        // disable the Fetch Data button
+        dispatch(simulatorActions.setPhaseInProgress(false));
+      }
+    } else {
+      toastr.error('Please see debug console for details.', 'WDC Validation Error');
+    }
+  };
+}
+
+export function handleDataCallback(tableName, data) {
+  return (dispatch, getState) => {
+    // if data is valid, add it to the appropriate tables object in the store
+    if (validateData(data)) {
+      const { tables } = getState();
+      let newTables = Object.assign({}, tables);
+      newTables[tableName].data = newTables[tableName].data.concat(data);
+      dispatch(simulatorActions.setTables(newTables));
+    } else {
+      toastr.error('Please see debug console for details.', 'WDC Validation Error');
+    }
+  };
+}
+
+export function handleDataDoneCallback() {
+  return (dispatch) => {
+    // Tell the wdc to shutdown now that we have our data
+    dispatch(simulatorActions.setPhaseInProgress(false));
+    dispatch(sendShutdown);
+  };
+}
+
+export function handleShutdownCallback() {
+  return (dispatch) => {
+    // Wdc has told us it's shutdown, end our phase
+    dispatch(simulatorActions.setPhaseInProgress(false));
+  };
+}
+
+export function handleAbort(errMsg) {
+  return (dispatch) => {
+    // Something went wrong, end phase, alert user
+    dispatch(simulatorActions.setPhaseInProgress(false));
+    console.error(errMsg);
+    toastr.error(errMsg, 'The WDC reported an error:');
+  };
+}
+
+export function handleAbortForAuth(errMsg) {
+  return (dispatch) => {
+    // Need auth, close the simulator, tell the user
+    dispatch(simulatorActions.setPhaseInProgress(false));
+    dispatch(simulatorActions.closeSimulator());
+    toastr.error(errMsg, 'The WDC has been aborted for auth, use the "Start Auth Phase" to test your WDC Auth Mode:');
+  };
+}
+
+// Send message thunks
+export function sendMessage(messageName, payload) {
+  return (dispatch, getState) => {
+    // Construct payload and send info to whichever
+    // window we are posting too
+    const { simulatorWindow, wdcAttrs } = getState();
+    const messagePayload = JSON.stringify({
+      msgName: messageName,
+      msgData: payload,
+      props: wdcAttrs,
+    });
+
+    if (simulatorWindow) {
```

Comment:
The only time I can think that we would send a message to an undefined window is if there is a race condition with when a user closes an interactive window manually but we are still sending/receiving messages to it. I can test to see if that is an issue and if not take that check out.